bob/src/cluster/quorum.rs

```diff
@@ -351,25 +352,194 @@ impl Cluster for Quorum {
+            }
+        }
+
+        if !keys_by_node.is_empty() {
+            let nodes: Vec<_> = keys_by_node.keys().cloned().collect();
+            let remote_results = exist_on_remote_nodes(&nodes, keys_by_node).await;
+            for (remote_result, node) in remote_results.into_iter().zip(nodes) {
+                match remote_result {
+                    Ok(remote_result) => {
+                        indexes_by_node
+                            .get(&node)
+                            .map(|idx| idx.update_existence(result, remote_result.inner()));
+                    }
+                    Err(e) => {
+                        debug!("Failed to check existence on node {}: {:?}", node.name(), e);
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn group_by_nodes(
+    keys: &[BobKey],
+    mapper: &Virtual,
+) -> (
+    Option<IndexMap>,
+    HashMap<Node, IndexMap>,
+    HashMap<Node, IndexMap>,
+) {
+    let mut local = IndexMap::new();
+    let mut primary = HashMap::new();
+    let mut secondary = HashMap::new();
+
+    let local_node = mapper.local_node_name();
+
+    for (index, &key) in keys.iter().enumerate() {
+        let target_nodes = mapper.get_target_nodes_for_key(key);
+
+        if !target_nodes
+            .iter()
+            .any(|n| n.name() == local_node || primary.contains_key(n))
+        {
+            if let Some(node) = target_nodes.iter().find(|n| !secondary.contains_key(*n)) {
+                primary.insert(node.clone(), IndexMap::new());
+            }
+        }
+
+        for node in target_nodes {
+            if node.name() == local_node {
+                local.push(index);
+            } else if let Some(map) = primary.get_mut(node) {
+                map.push(index)
+            } else {
+                secondary
+                    .entry(node.clone())
+                    .or_insert(IndexMap::new())
+                    .push(index);
+            }
+        }
+    }
+
+    (
+        if local.is_empty() { None } else { Some(local) },
+        primary,
+        secondary,
+    )
+}
+
+pub(crate) struct IndexMap {
+    indexes: Vec<usize>,
+}
+
+impl IndexMap {
+    pub(crate) fn new() -> Self {
+        Self { indexes: vec![] }
+    }
+
+    pub(crate) fn where_not_exists(data: &[bool]) -> Self {
+        Self {
+            indexes: data
+                .iter()
+                .enumerate()
+                .filter(|(_, f)| !**f)
+                .map(|(i, _)| i)
+                .collect(),
+        }
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.indexes.is_empty()
+    }
+
+    pub(crate) fn push(&mut self, item: usize) {
+        debug_assert!(!self.indexes.contains(&item));
+        self.indexes.push(item);
+    }
+
+    pub(crate) fn collect<'a, T: Clone + 'a>(
+        &'a self,
+        data: impl IntoIterator<Item = &'a T>,
+    ) -> Vec<T> {
+        data.into_iter()
```

Comment:
The complexity of implemented approach is `O(n*m)`. It is better to iterate over `indexes`, which will give the complexity `O(m)`. Also, please implement it with regular `for` loop to achieve the best possible performance. Additionally, the result vector can be pre-allocated with `Vec::with_capacity(self.indexes.len())`